src/gravity.js

```
@@ -0,0 +1,508 @@
+  var s = spriteData.slice();
+  if (v) {
+    for (y = 0; y < s.length / 2; ++y) {
+      y2 = s.length - y - 1;
+      tmp = s[y];
+      s[y] = s[y2];
+      s[y2] = tmp;
+    }
+  }
+  if (h) {
+    for (y = 0; y < s.length; ++y) {
+      col = s[y] = s[y].slice();
+      for (x = 0; x < col.length / 2; ++x) {
+        x2 = col.length - x - 1;
+        tmp = col[x];
+        col[x] = col[x2];
+        col[x2] = tmp;
+      }
+    }
+  }
+  if (rot) {
+    s = transpose(s)
+  }
+
+  return s;
+}
+
+function flipAvatar(gravityDirection) {
+  var hflip = false;
+  var vflip = false;
+  var rot = false;
+
+  var i;
+  // save the original frames
+  if (!originalAnimation || originalAnimation.referenceFrame !== getSpriteData(bitsy.playerId, 0)) {
+    originalAnimation = {
+      frames: []
+    };
+    for (i = 0; i < bitsy.player().animation.frameCount; ++i) {
+      originalAnimation.frames.push(getSpriteData(bitsy.playerId, i));
+    }
+  }
+
+  // determine which directions need flipping
+  switch (gravityDirection) {
+  case 'up':
+    vflip = true;
+    break;
+  case 'down':
+    vflip = false;
+    break;
+  case 'left':
+    vflip = true;
+    rot = true;
+    break;
+  case 'right':
+    rot = true;
+    break;
+  default:
+    break;
+  }
+
+  // update sprite with flipped frames
+  for (i = 0; i < originalAnimation.frames.length; ++i) {
+    setSpriteData(bitsy.playerId, i, flip(originalAnimation.frames[i], vflip, hflip, rot));
+  }
+  originalAnimation.referenceFrame = getSpriteData(bitsy.playerId, 0);
+}
+
+/* end of directional-avatar copy pasta */
+
+/* dialog tags
+*/
+
+
+addDualDialogTag("invertGravity", function (env, params) {
+  var newGravityDir = params[0];
+  invertPhysics(newGravityDir);
+  if (hackOptions.flipAvatarOnInversion) flipAvatar(newGravityDir);
+})
+
+addDualDialogTag('toggleGravity', function (env, params) {
+  active = !active;
+});
+
+addDualDialogTag('setJumpPower', function (env, params) {
+  var amt = ~~params[0];
```

hm, it would probably be better not to try to turn anything into a useable number and instead allow it to error (or catch and log the error manually) when given invalid input; having it silently behave as a zero will likely be more confusing to someone who made a mistake than an actual error

---

ooo i didn't know you could pass integers as params to dialog tags. cool! 
do you think i should add error handling here? Once i removed all the ~~'s and !!'s the code works whether someone passes a `1` or a `"1"` but if someone tries to pass `"one"` or something weird it's gonna yell about `NaN`

---

you could throw an error with an explicit message to make the invalid input clearer if you'd like, but that's the most that i'd do